src/syscalls.rs

```diff
@@ -97,19 +95,27 @@ pub unsafe fn command(major: usize, minor: usize, arg1: usize, arg2: usize) -> i
     res
 }
 
-pub unsafe fn allow(major: usize, minor: usize, slice: &[u8]) -> isize {
-    let res;
-    asm!("svc 3" : "={r0}"(res)
-                 : "{r0}"(major) "{r1}"(minor) "{r2}"(slice.as_ptr()) "{r3}"(slice.len())
-                 : "memory"
-                 : "volatile");
-    res
+pub fn allow(
+    driver_number: usize,
+    allow_number: usize,
+    buffer_to_share: &mut [u8],
+) -> Result<SharedMemory, isize> {
+    let return_code = unsafe { allow_ptr(driver_number, allow_number, buffer_to_share) };
+    if return_code == 0 {
+        Ok(SharedMemory {
+            driver_number,
+            allow_number,
+            buffer_to_share,
+        })
+    } else {
+        Err(return_code)
+    }
 }
 
-unsafe fn unallow(major: usize, minor: usize) -> isize {
+pub unsafe fn allow_ptr(major: usize, minor: usize, slice: &[u8]) -> isize {
```

Comment:
Since this function also accepts null pointers, the reference in `slice: &[u8]` might be misleading. In my opinion, the signature should be `... buffer: *mut [u8], length: usize ...`.